File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing/OptionsProvider.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Collections.Generic;
using Microsoft.CodeAnalysis.Diagnostics;

namespace Microsoft.CodeAnalysis.CSharp.Testing
{
    /// <summary>
    /// This class just passes argument through to the projects options provider and it used to provider custom global options
    /// </summary>
    internal class OptionsProvider : AnalyzerConfigOptionsProvider
    {
        private readonly AnalyzerConfigOptionsProvider _analyzerConfigOptionsProvider;

        [System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1414:Tuple types in signatures should have element names", Justification = "The tuple names in the global options are unknown.")]
        public OptionsProvider(AnalyzerConfigOptionsProvider analyzerConfigOptionsProvider, List<(string, string)> globalOptions)
        {
            _analyzerConfigOptionsProvider = analyzerConfigOptionsProvider;
            GlobalOptions = new ConfigOptions(_analyzerConfigOptionsProvider.GlobalOptions, globalOptions);
        }

        public override AnalyzerConfigOptions GlobalOptions { get; }

        public override AnalyzerConfigOptions GetOptions(SyntaxTree tree)
            => _analyzerConfigOptionsProvider.GetOptions(tree);

        public override AnalyzerConfigOptions GetOptions(AdditionalText textFile)
            => _analyzerConfigOptionsProvider.GetOptions(textFile);
    }
}

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.VisualBasic.SourceGenerators.Testing/OptionsProvider.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports Microsoft.CodeAnalysis.Diagnostics

Friend Class OptionsProvider
    Inherits AnalyzerConfigOptionsProvider

    Private ReadOnly _analyzerConfigOptionsProvider As AnalyzerConfigOptionsProvider
    Private ReadOnly _globalOptions As AnalyzerConfigOptions

    Public Sub New(analyzerConfigOptionsProvider As AnalyzerConfigOptionsProvider, globalOptions As List(Of (String, String)))
        _analyzerConfigOptionsProvider = analyzerConfigOptionsProvider
        _globalOptions = New ConfigOptions(_analyzerConfigOptionsProvider.GlobalOptions, globalOptions)
    End Sub

    Public Overrides ReadOnly Property GlobalOptions As AnalyzerConfigOptions
        Get
            Return _globalOptions
        End Get
    End Property

    Public Overrides Function GetOptions(tree As SyntaxTree) As AnalyzerConfigOptions
        Return _analyzerConfigOptionsProvider.GetOptions(tree)
    End Function

    Public Overrides Function GetOptions(additionalFile As AdditionalText) As AnalyzerConfigOptions
        Return _analyzerConfigOptionsProvider.GetOptions(additionalFile)
    End Function

End Class
